File: cryptography/RSA/c/rsa.c
// C program for RSA asymmetric cryptographic
// algorithm. For demonstration values are relatively
// small compared to practical application
#include<stdio.h>
#include<math.h>
 
// Returns gcd of a and b
int gcd(int a, int b)
{
    if (a == 0)
        return b;
    return gcd(b%a, a);
}
 
// Code to demonstrate RSA algorithm
int main()
{
    // Two random prime numbers
    double p = 3;
    double q = 7;
 
    // First part of public key:
    double n = p*q;
 
    // Finding other part of public key.
    // e stands for encrypt
    double e = 2;
    double phi = (p-1)*(q-1);
    while (e < phi)
    {
        // e must be co-prime to phi and
        // smaller than phi.
        if (gcd(e, phi)==1)
            break;
        else
            e++;
    }
 
    // Private key (d stands for decrypt)
    int k = 2;  // A constant value
    double d = (1 + (k*phi))/e;
 
    // Message to be encrypted
    double msg = 12;
    printf("Message data = %lf",msg);
 
    // Encryption c = (msg ^ e) % n
    double c = pow(msg, e);
    c = fmod(c, n);
    printf("\nEncrypted data = %lf", c);
 
    // Decryption m = (c ^ d) % n
    double m = pow(c, d);
    m = fmod(m, n);
    printf("\nOriginal Message Sent = %lf",m);
 
    return 0;
}
// This code is contributed by Akash Sharan.

File: cryptography/RSA/golang/rsa.go
// RSA implementation using bigInt
package main

import (
	"fmt"
	"math/big"
)

// Code to demonstrate RSA algorithm
func main() {
	// Defined big.Int in value: 1
	one := new(big.Int).SetInt64(1)

	// Two random prime numbers
	p := new(big.Int).SetInt64(3)
	q := new(big.Int).SetInt64(7)

	// n = pq
	n := new(big.Int).Mul(p, q)

	// Message
	msg := new(big.Int).SetInt64(12)

	// Start finding other part of public key
	// 1 < e < phi(n)
	e := new(big.Int).SetInt64(2)

	// phi(n)
	phi := new(big.Int).Mul(new(big.Int).Sub(p, one), new(big.Int).Sub(q, one))

	// Find e, which GCD(e, phi) == 1, e starts from 2
	for e.Cmp(phi) == -1 {
		if new(big.Int).GCD(nil, nil, e, phi).Cmp(one) == 0 {
			break
		} else {
			e.Add(e, one)
		}
	}

	// Private key
	d := e
	d.ModInverse(d, phi)

	// Encrypted
	c := new(big.Int).Exp(msg, e, n)
	// Decrypted
	m := new(big.Int).Exp(c, d, n)

	fmt.Printf("Message data = %d\n", msg)
	fmt.Printf("Encrypted data = %d\n", c)
	fmt.Printf("Original Message Sent = %d\n", m)
}
